### PR TITLE
8283326: Implement SafeFetch statically

### PR DESCRIPTION
**Pull request by snazarkin**
Hi! 
I'd like to backport this changeset to prevent a crash when AsyncProfiler is used on Mac M1. 

The backport is not clean:
1. RISCV is not part of jdk17
2. mallocTracker.cpp is not updated since 8281023 was not backported to jdk17
3. vmError_posix.cpp requires manual update due to  8289350 is missed at jdk19+
4. Need to add globals.hpp to fix builds without precompiled headers

There is follow up for this case: 
1. 8284997: arm32 build crashes since JDK-8283326
2. 8288719: [arm32] SafeFetch32 thumb interleaving causes random crashes

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8283326](https://bugs.openjdk.org/browse/JDK-8283326): Implement SafeFetch statically (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/1557/head:pull/1557` \
`$ git checkout pull/1557`

Update a local copy of the PR: \
`$ git checkout pull/1557` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/1557/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1557`

View PR using the GUI difftool: \
`$ git pr show -t 1557`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/1557.diff">https://git.openjdk.org/jdk17u-dev/pull/1557.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/1557#issuecomment-1625199354)

### CODE REVIEW

**Comment by VladimirKempik** [comment]
>The backport is not clean:

>RISCV is not part of jdk17

Hello Sergey
Could I ask you to wait a little bit until this ( https://github.com/openjdk/jdk17u-dev/pull/1427 ) integrated ? 

then you can include risc-v part into your backport and we won't have to backport risc-v of this patch separately

**Comment by tstuefe** [comment]
> > The backport is not clean:
> 
> > RISCV is not part of jdk17
> 
> Hello Sergey Could I ask you to wait a little bit until this ( #1427 ) integrated ?
> 
> then you can include risc-v part into your backport and we won't have to backport risc-v part of this patch separately

+1

It would also give me time to review this backport - I'm a bit snowed in atm but would like to take a closer look.

**Comment by snazarkin** [comment]
> > The backport is not clean:
> 
> > RISCV is not part of jdk17
> 
> Hello Sergey Could I ask you to wait a little bit until this ( #1427 ) integrated ?
> 
> then you can include risc-v part into your backport and we won't have to backport risc-v part of this patch separately

OK, I'll wait

**Comment by GoeLin** [comment]
RiskV will soon be pushed to 17. Please wait for this and include the RiskV parts.

File: src/hotspot/os/posix/vmError_posix.cpp [hunk 1]
```diff
@@ -76,22 +75,19 @@ static void crash_handler(int sig, siginfo_t* info, void* ucVoid) {
...
-      if (handle_assert_poison_fault(ucVoid, info->si_addr)) {
-        return;
-      }
+  if ((sig == SIGSEGV || sig == SIGBUS) && info != NULL && info->si_addr == g_assert_poison) {
+    if (handle_assert_poison_fault(ucVoid, info->si_addr)) {
+      return;
     }
-#endif // CAN_SHOW_REGISTERS_ON_ASSERT
   }
+#endif // CAN_SHOW_REGISTERS_ON_ASSERT
```
Comment (tstuefe) [inline src/hotspot/os/posix/vmError_posix.cpp:90]: Above hunks are not part of the patch, and should not be needed for safefetch. Why are they here?

**Review by tstuefe — COMMENTED** [review]
Looks good, apart from one note and the missing riscv parts.

**Comment by VladimirKempik** [comment]
risc-v port is in.

**Comment by GoeLin** [comment]
Hi, please first get the PR ready, then label fix-request

File: src/hotspot/os/posix/vmError_posix.cpp (hunk 1 above)
Comment (snazarkin) [reply src/hotspot/os/posix/vmError_posix.cpp:90]: It's here to keep existing logic of this hunk. And to sync with mainline JDK version. The file was previously modified in scope of JDK-8289350 that doesn't affect JDK19+.

**Comment by snazarkin** [comment]
Rebased after RISCV integration, although all comments should remain in place

File: src/hotspot/os/posix/vmError_posix.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/posix/vmError_posix.cpp:90]: Okay, I see now.

**Review by tstuefe — CHANGES_REQUESTED** [review]
riscv parts are still missing.

Would be good to get a test from @RealFYang on linux riscv (gtest should be enough).

**Comment by RealFYang** [comment]
> riscv parts are still missing.
> 
> Would be good to get a test from @RealFYang on linux riscv (gtest should be enough).

@tstuefe : I did a native build on my new 64-core Milk-V Pioneer box and run make test TEST="gtest:all". gtest result is clean.

**Review by tstuefe — APPROVED** [review]
Looks good.

When you do the fix request, please mention the two addon changes needed for arm32. I don't remember the procedure for downporting such change chains, but @GoeLin will know.

**Comment by tstuefe** [comment]
> > riscv parts are still missing.
> > Would be good to get a test from @RealFYang on linux riscv (gtest should be enough).
> 
> @tstuefe : I did a native build on my new 64-core Milk-V Pioneer box and run make test TEST="gtest:all". gtest result is clean.

Thank you @RealFYang and also thats a nice name for a processor.

**Comment by snazarkin** [comment]
Many thanks to @tstuefe  and @RealFYang.

**Comment by snazarkin** [comment]
Got the push approval.

/integrate
